cmd/pow/cmd/ffs_stage.go

```diff
@@ -31,18 +39,84 @@ var ffsStageCmd = &cobra.Command{
 		defer cancel()
 
 		if len(args) != 1 {
-			Fatal(errors.New("you must provide a file path"))
+			Fatal(errors.New("you must provide a file/folder path"))
 		}
 
-		f, err := os.Open(args[0])
-		checkErr(err)
-		defer func() { checkErr(f.Close()) }()
+		fi, err := os.Stat(args[0])
+		if os.IsNotExist(err) {
+			Fatal(errors.New("file/folder doesn't exist"))
+		}
+		if err != nil {
+			Fatal(fmt.Errorf("getting file/folder information: %s", err))
+		}
+		var cid cid.Cid
+		if fi.IsDir() {
+			token := viper.GetString("token")
+			cid, err = addFolder(args[0], token)
+			if err != nil {
+				Fatal(fmt.Errorf("staging folder: %s", err))
+			}
+		} else {
+			f, err := os.Open(args[0])
+			checkErr(err)
+			defer func() { checkErr(f.Close()) }()
 
-		s := spin.New("%s Caching specified file in FFS hot storage...")
-		s.Start()
-		cid, err := fcClient.FFS.Stage(authCtx(ctx), f)
-		s.Stop()
-		checkErr(err)
-		Success("Cached file in FFS hot storage with cid: %s", util.CidToString(*cid))
+			s := spin.New("%s Staging specified file in FFS hot storage...")
+			s.Start()
+			ptrCid, err := fcClient.FFS.Stage(authCtx(ctx), f)
+			s.Stop()
+			checkErr(err)
+			cid = *ptrCid
+		}
+		Success("Staged asset in FFS hot storage with cid: %s", util.CidToString(cid))
 	},
 }
+
+func addFolder(folderPath string, ffsToken string) (cid.Cid, error) {
+	ipfsProxy := viper.GetString("ipfsrevproxy")
+	ma, _ := multiaddr.NewMultiaddr(ipfsProxy)
+	customClient := http.DefaultClient
+	customClient.Transport = newFFSHeaderDecorator(ffsToken)
+	ipfs, err := httpapi.NewApiWithClient(ma, customClient)
+	if err != nil {
+		return cid.Undef, err
+	}
+	stat, err := os.Lstat(folderPath)
+	if err != nil {
+		return cid.Undef, err
+	}
+	ff, err := files.NewSerialFile(folderPath, false, stat)
+	if err != nil {
+		return cid.Undef, err
+	}
+	defer func() {
+		if err := ff.Close(); err != nil {
+			Fatal(fmt.Errorf("closing folder: %s", err))
+		}
+	}()
+	opts := []options.UnixfsAddOption{
+		options.Unixfs.CidVersion(1),
+		options.Unixfs.Pin(false),
+	}
+	pth, err := ipfs.Unixfs().Add(context.Background(), files.ToDir(ff), opts...)
+	if err != nil {
+		return cid.Undef, err
+	}
+	return pth.Cid(), nil
+}
+
+type ffsHeaderDecorator struct {
+	ffsToken string
+}
+
+func newFFSHeaderDecorator(ffsToken string) *ffsHeaderDecorator {
+	return &ffsHeaderDecorator{
+		ffsToken: ffsToken,
+	}
+}
+
+func (fhd ffsHeaderDecorator) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.Header["x-ipfs-ffs-auth"] = []string{fhd.ffsToken}
+
+	return http.DefaultTransport.RoundTrip(req)
+}
```

Comment:
All http calls made by the IPFS client lib, will be hooked to add an auth header to be authenticated.
The server-side reverse proxy checks for this header, before redirecting the request to  the IPFS node.